parlai/core/gpt2_helper.py

```diff
@@ -0,0 +1,161 @@
+
+
+@lru_cache()
+def bytes_to_unicode():
+    """
+    Returns list of utf-8 byte and a corresponding list of unicode strings.
+    The reversible bpe codes work on unicode strings.
+    This means you need a large # of unicode characters in your vocab if you want to avoid UNKs.
+    When you're at something like a 10B token dataset you end up needing around 5K for decent coverage.
+    This is a signficant percentage of your normal, say, 32K bpe vocab.
+    To avoid that, we want lookup tables between utf-8 bytes and unicode strings.
+    And avoids mapping to whitespace/control characters the bpe code barfs on.
+    """
+    bs = (
+        list(range(ord("!"), ord("~") + 1))
+        + list(range(ord("¡"), ord("¬") + 1))
+        + list(range(ord("®"), ord("ÿ") + 1))
+    )
+    cs = bs[:]
+    n = 0
+    for b in range(2 ** 8):
+        if b not in bs:
+            bs.append(b)
+            cs.append(2 ** 8 + n)
+            n += 1
+    cs = [chr(n) for n in cs]
+    return dict(zip(bs, cs))
+
+
+def get_pairs(word):
+    """Return set of symbol pairs in a word.
+    Word is represented as tuple of symbols (symbols being variable-length strings).
+    """
+    pairs = set()
+    prev_char = word[0]
+    for char in word[1:]:
+        pairs.add((prev_char, char))
+        prev_char = char
+    return pairs
+
+
+class Gpt2BpeHelper:
+    def __init__(self, opt, errors='replace'):
+        data_path = os.path.join(opt['datapath'], 'gpt2')
+        vocab_path = os.path.join(data_path, 'vocab.bpe')
+        json_path = os.path.join(data_path, 'encoder.json')
+        if not os.path.isfile(vocab_path) or not os.path.isfile(json_path):
+            make_dir(data_path)
+            download(DEFAULT_VOCAB_BPE, data_path, 'vocab.bpe')
+            download(DEFAULT_ENCODER_JSON, data_path, 'encoder.json')
+        with open(vocab_path, 'r', encoding="utf-8") as f:
+            bpe_data = f.read()
+        with open(json_path, 'r') as f:
+            self.encoder = json.load(f)
+        for each_token in self.encoder.keys():
+            new_token = ''.join(
+                # escape nonprintable characters
+                '\\' + hex(b).lstrip('0') if (b > 127 or b < 32) else chr(b)
+                for b in each_token.encode('utf-8')
+            )
+            self.encoder[each_token] = new_token
+        self.decoder = {v: k for k, v in self.encoder.items()}
+        bpe_merges = [
+            tuple(merge_str.split()) for merge_str in bpe_data.split('\n')[1:-1]
+        ]
+        self.errors = errors  # how to handle errors in decoding
+        self.byte_encoder = bytes_to_unicode()
+        self.byte_decoder = {v: k for k, v in self.byte_encoder.items()}
+        self.bpe_ranks = dict(zip(bpe_merges, range(len(bpe_merges))))
+        self.cache = {}
+        try:
+            import regex as re
+
+            self.re = re
+        except ImportError:
+            raise ImportError('Please install regex with: pip install regex')
```

Comment:
is this not a requirement in parlai already? seems like it should be if not

Reply:
This was copy pasta from fairseq.